File: 0201-9-3045-count-prefix-and-suffix-pairs-ii/0201-9-3045-count-prefix-and-suffix-pairs-ii.py
class TrieNode:
    def __init__(self):
        self.child = {}
        self.frq = 0

class Solution:
    def __init__(self):
        self.trie = TrieNode()

    def insert(self, s):
        top = self.trie
        n = len(s)
        for i in range(n):
            pre = s[i]
            suf = s[n - i - 1]
            if (pre, suf) not in top.child:
                top.child[(pre, suf)] = TrieNode()
            top = top.child[(pre, suf)]
            top.frq += 1

    def search(self, s):
        top = self.trie
        n = len(s)
        for i in range(n):
            pre = s[i]
            suf = s[n - i - 1]
            if (pre, suf) not in top.child:
                return 0
            top = top.child[(pre, suf)]
        return top.frq

    def countPrefixSuffixPairs(self, words):
        cnt = 0
        for word in reversed(words):
            cnt += self.search(word)
            self.insert(word)
        return cnt

File: 0201-9-3045-count-prefix-and-suffix-pairs-ii/0201-9-3045-count-prefix-and-suffix-pairs-ii.cpp
struct TrieNode{
    map<pair<char,char>,TrieNode*>child;
    int frq = 0;
}*trie;
class Solution {
public:
    void insert(string s){
        TrieNode* top = trie;
        int n = s.size();
        for(int i = 0 ;  i < n;i++){
            char pre = s[i], suf = s[n-i-1];
            if(!top->child[{pre,suf}]) top->child[{pre,suf}] = new TrieNode();
            top = top->child[{pre,suf}];
            top->frq++;
        }
    }
    long long search(string s){
        TrieNode* top = trie;
        int n = s.size();
        for(int i = 0 ;  i < n;i++){
            char pre = s[i], suf = s[n-i-1];
            if(!top->child[{pre,suf}]) return 0;
            top = top->child[{pre,suf}];
        }
        return top->frq;
    }
    long long countPrefixSuffixPairs(vector<string>& words) {
        long long cnt = 0;
        trie = new TrieNode();
        for(int i = words.size() - 1 ; i >= 0 ; i--){
            cnt+=search(words[i]);
            insert(words[i]);
        }
        return cnt;

    }
};